### ISSUE TITLE
Outbound proxy routes traffic to wrong resp. non-longer existing IP addresses

### ISSUE THREAD

**Issue by rustrial** [linkerd/linkerd2#6129]
## Bug Report

We observe a strange behaviour using linkerd version `edge-21.4.3`. The outbound `linkerd-proxy` sidecars of our (gloo-edge) ingress gateway instances (running in [ingress mode](https://linkerd.io/2.10/tasks/using-ingress/)) sometimes send traffic to the wrong IP addresses. And if those IP addresses have been reassigned to other Pods or EC2 instances (in the same VPC) such traffic is routed to the wrong Pods resp. EC2 instances.

Note, this is a serious issue as sending data to the wrong IP address is a **potential data breach** as it might be forwarded to a Pod in a different Kubernetes namespace or any other network interface in the same AWS VPC that is now using that IP address.

Restarting the outbound `linkerd-proxy` in our gloo-edge ingress gateway Pods resolves the problem. So it looks like this happens due to some bad state inside the outbound `linkerd-proxy`.

### What is the issue?

The outbound `linkerd-proxy` is forwarding traffic to wrong (stale) IP addresses.

### How can it be reproduced?

We cannot reproduce this behaviour. Currently we have a production system affected and in order to being able to investigate this issue, we have not yet restarted all `linkerd-proxy` containers (of our gloo-edge Pods) in that environment.

### Logs, error output, etc

Bellow you can find a log excerpt from the outbound `linkerd-proxy` in `debug` mode over a 1 second time window (without any filtering).

- The following IP addresses `10.2.36.185, 10.2.1.194, 10.2.0.179, 10.2.14.66` used to connect to identity `gh-commander.nx-devices.serviceaccount.identity.linkerd.cluster.local` are no longer valid (there are no Pods assigned to those addresses). 
- The IP address `10.2.45.164` is assigned to a Pod called `gh-mgmt-service-kafka-consumer-8559dc854d-chs9n`, which is not part of any `Service` and is **not** related to the identity `gh-commander.nx-devices.serviceaccount.identity.linkerd.cluster.local` at all. I guess this IP address has initially been used by a Pod of `gh-commander.nx-devices.serviceaccount.identity.linkerd.cluster.local` and later (as that Pod terminated) been reassigned to another (unrelated) Pod.

So it looks like the outbound proxy has some very stale cache entries which should have been removed long ago.

```
2021-05-13T16:36:12+02:00 {} {"log":"{"timestamp":"[2066725.714142s]","level":"DEBUG","fields":{"message":"Building HTTP client","settings":"OrigProtoUpgrade"},"target":"linkerd_proxy_http::client","spans":[{"orig_dst":"10.2.45.164:8000","name":"ingress"},{"dst":"10.2.45.164:8000","name":"target"}],"threadId":"ThreadId(2)"}\n","stream":"stdout","time":"2021-05-13T14:36:11.811526782Z"}
2021-05-13T16:36:12+02:00 {} {"log":"{"timestamp":"[2066725.714210s]","level":"DEBUG","fields":{"message":"Initiating TLS connection","server.id":"gh-commander.nx-devices.serviceaccount.identity.linkerd.cluster.local"},"target":"linkerd_tls::client","spans":[{"orig_dst":"10.2.45.164:8000","name":"ingress"},{"dst":"10.2.45.164:8000","name":"target"}],"threadId":"ThreadId(2)"}\n","stream":"stdout","time":"2021-05-13T14:36:11.811571433Z"}
2021-05-13T16:36:12+02:00 {} {"log":"{"timestamp":"[2066725.714224s]","level":"DEBUG","fields":{"message":"Connecting","server.addr":"10.2.45.164:8000"},"target":"linkerd_proxy_transport::connect","spans":[{"orig_dst":"10.2.45.164:8000","name":"ingress"},{"dst":"10.2.45.164:8000","name":"target"}],"threadId":"ThreadId(2)"}\n","stream":"stdout","time":"2021-05-13T14:36:11.811589523Z"}
2021-05-13T16:36:12+02:00 {} {"log":"{"timestamp":"[2066725.715531s]","level":"DEBUG","fields":{"message":"Connected","local.addr":"10.2.17.237:60492","keepalive":"Some(10s)"},"target":"linkerd_proxy_transport::connect","spans":[{"orig_dst":"10.2.45.164:8000","name":"ingress"},{"dst":"10.2.45.164:8000","name":"target"},{"name":"h2"}],"threadId":"ThreadId(3)"}\n","stream":"stdout","time":"2021-05-13T14:36:11.812795897Z"}
2021-05-13T16:36:12+02:00 {} {"log":"{"timestamp":"[2066725.718905s]","level":"DEBUG","fields":{"message":"Failed to connect","error":"received corrupt message"},"target":"linkerd_reconnect::service","spans":[{"orig_dst":"10.2.45.164:8000","name":"ingress"},{"dst":"10.2.45.164:8000","name":"target"}],"threadId":"ThreadId(2)"}\n","stream":"stdout","time":"2021-05-13T14:36:11.816252567Z"}
2021-05-13T16:36:12+02:00 {} {"log":"{"timestamp":"[2066725.718950s]","level":"DEBUG","fields":{"message":"Recovering"},"target":"linkerd_reconnect::service","spans":[{"orig_dst":"10.2.45.164:8000","name":"ingress"},{"dst":"10.2.45.164:8000","name":"target"}],"threadId":"ThreadId(2)"}\n","stream":"stdout","time":"2021-05-13T14:36:11.816289167Z"}
2021-05-13T16:36:12+02:00 {} {"log":"{"timestamp":"[2066725.755458s]","level":"DEBUG","fields":{"method":"POST","uri":"http://rest_xds_cluster/v3/discovery:endpoints","version":"HTTP/1.1"},"target":"linkerd_proxy_http::client","spans":[{"orig_dst":"10.2.11.155:9976","name":"ingress"},{"v":"1.x","name":"http"},{"dst":"10.2.11.155:9976","name":"target"},{"server.addr":"10.2.11.155:9976","name":"endpoint"},{"name":"orig-proto-upgrade"}],"threadId":"ThreadId(3)"}\n","stream":"stdout","time":"2021-05-13T14:36:11.853131817Z"}
2021-05-13T16:36:12+02:00 {} {"log":"{"timestamp":"[2066725.755509s]","level":"DEBUG","fields":{"headers":"{"host": "rest_xds_cluster", "content-type": "application/json", "content-length": "14661", "x-envoy-internal": "true", "x-forwarded-for": "10.2.17.237", "x-envoy-expected-rq-timeout-ms": "5000"}"},"target":"linkerd_proxy_http::client","spans":[{"orig_dst":"10.2.11.155:9976","name":"ingress"},{"v":"1.x","name":"http"},{"dst":"10.2.11.155:9976","name":"target"},{"server.addr":"10.2.11.155:9976","name":"endpoint"},{"name":"orig-proto-upgrade"}],"threadId":"ThreadId(3)"}\n","stream":"stdout","time":"2021-05-13T14:36:11.853168717Z"}
2021-05-13T16:36:12+02:00 {} {"log":"{"timestamp":"[2066725.755539s]","level":"DEBUG","fields":{"message":"Upgrading request","version":"HTTP/1.1","absolute_form":false},"target":"linkerd_proxy_http::orig_proto","spans":[{"orig_dst":"10.2.11.155:9976","name":"ingress"},{"v":"1.x","name":"http"},{"dst":"10.2.11.155:9976","name":"target"},{"server.addr":"10.2.11.155:9976","name":"endpoint"},{"name":"orig-proto-upgrade"}],"threadId":"ThreadId(3)"}\n","stream":"stdout","time":"2021-05-13T14:36:11.853178028Z"}
2021-05-13T16:36:12+02:00 {} {"log":"{"timestamp":"[2066725.758007s]","level":"DEBUG","fields":{"message":"Building HTTP client","settings":"OrigProtoUpgrade"},"target":"linkerd_proxy_http::client","spans":[{"orig_dst":"10.2.36.185:8000","name":"ingress"},{"dst":"10.2.36.185:8000","name":"target"}],"threadId":"ThreadId(2)"}\n","stream":"stdout","time":"2021-05-13T14:36:11.855422573Z"}
2021-05-13T16:36:12+02:00 {} {"log":"{"timestamp":"[2066725.758061s]","level":"DEBUG","fields":{"message":"Initiating TLS connection","server.id":"gh-commander.nx-devices.serviceaccount.identity.linkerd.cluster.local"},"target":"linkerd_tls::client","spans":[{"orig_dst":"10.2.36.185:8000","name":"ingress"},{"dst":"10.2.36.185:8000","name":"target"}],"threadId":"ThreadId(2)"}\n","stream":"stdout","time":"2021-05-13T14:36:11.855439903Z"}
2021-05-13T16:36:12+02:00 {} {"log":"{"timestamp":"[2066725.758076s]","level":"DEBUG","fields":{"message":"Connecting","server.addr":"10.2.36.185:8000"},"target":"linkerd_proxy_transport::connect","spans":[{"orig_dst":"10.2.36.185:8000","name":"ingress"},{"dst":"10.2.36.185:8000","name":"target"}],"threadId":"ThreadId(2)"}\n","stream":"stdout","time":"2021-05-13T14:36:11.855445123Z"}
2021-05-13T16:36:12+02:00 {} {"log":"{"timestamp":"[2066725.836934s]","level":"DEBUG","fields":{"message":"Building HTTP client","settings":"OrigProtoUpgrade"},"target":"linkerd_proxy_http::client","spans":[{"orig_dst":"10.2.14.66:8000","name":"ingress"},{"dst":"10.2.14.66:8000","name":"target"}],"threadId":"ThreadId(3)"}\n","stream":"stdout","time":"2021-05-13T14:36:11.934257073Z"}
2021-05-13T16:36:12+02:00 {} {"log":"{"timestamp":"[2066725.837009s]","level":"DEBUG","fields":{"message":"Initiating TLS connection","server.id":"gh-commander.nx-devices.serviceaccount.identity.linkerd.cluster.local"},"target":"linkerd_tls::client","spans":[{"orig_dst":"10.2.14.66:8000","name":"ingress"},{"dst":"10.2.14.66:8000","name":"target"}],"threadId":"ThreadId(3)"}\n","stream":"stdout","time":"2021-05-13T14:36:11.934306093Z"}
2021-05-13T16:36:12+02:00 {} {"log":"{"timestamp":"[2066725.837027s]","level":"DEBUG","fields":{"message":"Connecting","server.addr":"10.2.14.66:8000"},"target":"linkerd_proxy_transport::connect","spans":[{"orig_dst":"10.2.14.66:8000","name":"ingress"},{"dst":"10.2.14.66:8000","name":"target"}],"threadId":"ThreadId(3)"}\n","stream":"stdout","time":"2021-05-13T14:36:11.934540986Z"}
2021-05-13T16:36:12+02:00 {} {"log":"{"timestamp":"[2066725.855221s]","level":"DEBUG","fields":{"method":"POST","uri":"http://rest_xds_cluster/v3/discovery:endpoints","version":"HTTP/1.1"},"target":"linkerd_proxy_http::client","spans":[{"orig_dst":"10.2.11.155:9976","name":"ingress"},{"v":"1.x","name":"http"},{"dst":"10.2.11.155:9976","name":"target"},{"server.addr":"10.2.11.155:9976","name":"endpoint"},{"name":"orig-proto-upgrade"}],"threadId":"ThreadId(2)"}\n","stream":"stdout","time":"2021-05-13T14:36:11.952562021Z"}
2021-05-13T16:36:12+02:00 {} {"log":"{"timestamp":"[2066725.855263s]","level":"DEBUG","fields":{"headers":"{"host": "rest_xds_cluster", "content-type": "application/json", "content-length": "14677", "x-envoy-internal": "true", "x-forwarded-for": "10.2.17.237", "x-envoy-expected-rq-timeout-ms": "5000"}"},"target":"linkerd_proxy_http::client","spans":[{"orig_dst":"10.2.11.155:9976","name":"ingress"},{"v":"1.x","name":"http"},{"dst":"10.2.11.155:9976","name":"target"},{"server.addr":"10.2.11.155:9976","name":"endpoint"},{"name":"orig-proto-upgrade"}],"threadId":"ThreadId(2)"}\n","stream":"stdout","time":"2021-05-13T14:36:11.952593412Z"}
2021-05-13T16:36:12+02:00 {} {"log":"{"timestamp":"[2066725.855286s]","level":"DEBUG","fields":{"message":"Upgrading request","version":"HTTP/1.1","absolute_form":false},"target":"linkerd_proxy_http::orig_proto","spans":[{"orig_dst":"10.2.11.155:9976","name":"ingress"},{"v":"1.x","name":"http"},{"dst":"10.2.11.155:9976","name":"target"},{"server.addr":"10.2.11.155:9976","name":"endpoint"},{"name":"orig-proto-upgrade"}],"threadId":"ThreadId(2)"}\n","stream":"stdout","time":"2021-05-13T14:36:11.952601102Z"}
2021-05-13T16:36:12+02:00 {} {"log":"{"timestamp":"[2066725.870690s]","level":"DEBUG","fields":{"message":"Building HTTP client","settings":"OrigProtoUpgrade"},"target":"linkerd_proxy_http::client","spans":[{"orig_dst":"10.2.1.194:8000","name":"ingress"},{"dst":"10.2.1.194:8000","name":"target"}],"threadId":"ThreadId(2)"}\n","stream":"stdout","time":"2021-05-13T14:36:11.968036728Z"}
2021-05-13T16:36:12+02:00 {} {"log":"{"timestamp":"[2066725.870755s]","level":"DEBUG","fields":{"message":"Initiating TLS connection","server.id":"gh-commander.nx-devices.serviceaccount.identity.linkerd.cluster.local"},"target":"linkerd_tls::client","spans":[{"orig_dst":"10.2.1.194:8000","name":"ingress"},{"dst":"10.2.1.194:8000","name":"target"}],"threadId":"ThreadId(2)"}\n","stream":"stdout","time":"2021-05-13T14:36:11.968072418Z"}
2021-05-13T16:36:12+02:00 {} {"log":"{"timestamp":"[2066725.870770s]","level":"DEBUG","fields":{"message":"Connecting","server.addr":"10.2.1.194:8000"},"target":"linkerd_proxy_transport::connect","spans":[{"orig_dst":"10.2.1.194:8000","name":"ingress"},{"dst":"10.2.1.194:8000","name":"target"}],"threadId":"ThreadId(2)"}\n","stream":"stdout","time":"2021-05-13T14:36:11.968080648Z"}
2021-05-13T16:36:12+02:00 {} {"log":"{"timestamp":"[2066725.935636s]","level":"DEBUG","fields":{"method":"POST","uri":"http://rest_xds_cluster/v3/discovery:endpoints","version":"HTTP/1.1"},"target":"linkerd_proxy_http::client","spans":[{"orig_dst":"10.2.11.155:9976","name":"ingress"},{"v":"1.x","name":"http"},{"dst":"10.2.11.155:9976","name":"target"},{"server.addr":"10.2.11.155:9976","name":"endpoint"},{"name":"orig-proto-upgrade"}],"threadId":"ThreadId(3)"}\n","stream":"stdout","time":"2021-05-13T14:36:12.032974049Z"}
2021-05-13T16:36:12+02:00 {} {"log":"{"timestamp":"[2066725.935687s]","level":"DEBUG","fields":{"headers":"{"host": "rest_xds_cluster", "content-type": "application/json", "content-length": "14660", "x-envoy-internal": "true", "x-forwarded-for": "10.2.17.237", "x-envoy-expected-rq-timeout-ms": "5000"}"},"target":"linkerd_proxy_http::client","spans":[{"orig_dst":"10.2.11.155:9976","name":"ingress"},{"v":"1.x","name":"http"},{"dst":"10.2.11.155:9976","name":"target"},{"server.addr":"10.2.11.155:9976","name":"endpoint"},{"name":"orig-proto-upgrade"}],"threadId":"ThreadId(3)"}\n","stream":"stdout","time":"2021-05-13T14:36:12.033019889Z"}
2021-05-13T16:36:12+02:00 {} {"log":"{"timestamp":"[2066725.935718s]","level":"DEBUG","fields":{"message":"Upgrading request","version":"HTTP/1.1","absolute_form":false},"target":"linkerd_proxy_http::orig_proto","spans":[{"orig_dst":"10.2.11.155:9976","name":"ingress"},{"v":"1.x","name":"http"},{"dst":"10.2.11.155:9976","name":"target"},{"server.addr":"10.2.11.155:9976","name":"endpoint"},{"name":"orig-proto-upgrade"}],"threadId":"ThreadId(3)"}\n","stream":"stdout","time":"2021-05-13T14:36:12.03303246Z"}
2021-05-13T16:36:12+02:00 {} {"log":"{"timestamp":"[2066725.982635s]","level":"DEBUG","fields":{"message":"Failed to connect","error":"request timed out"},"target":"linkerd_reconnect::service","spans":[{"orig_dst":"10.2.17.31:8000","name":"ingress"},{"dst":"10.2.17.31:8000","name":"target"}],"threadId":"ThreadId(2)"}\n","stream":"stdout","time":"2021-05-13T14:36:12.079951255Z"}
2021-05-13T16:36:12+02:00 {} {"log":"{"timestamp":"[2066725.982685s]","level":"DEBUG","fields":{"message":"Recovering"},"target":"linkerd_reconnect::service","spans":[{"orig_dst":"10.2.17.31:8000","name":"ingress"},{"dst":"10.2.17.31:8000","name":"target"}],"threadId":"ThreadId(2)"}\n","stream":"stdout","time":"2021-05-13T14:36:12.079985065Z"}
2021-05-13T16:36:12+02:00 {} {"log":"{"timestamp":"[2066726.011366s]","level":"DEBUG","fields":{"message":"Building HTTP client","settings":"OrigProtoUpgrade"},"target":"linkerd_proxy_http::client","spans":[{"orig_dst":"10.2.0.179:8000","name":"ingress"},{"dst":"10.2.0.179:8000","name":"target"}],"threadId":"ThreadId(2)"}\n","stream":"stdout","time":"2021-05-13T14:36:12.108727663Z"}
2021-05-13T16:36:12+02:00 {} {"log":"{"timestamp":"[2066726.011429s]","level":"DEBUG","fields":{"message":"Initiating TLS connection","server.id":"gh-commander.nx-devices.serviceaccount.identity.linkerd.cluster.local"},"target":"linkerd_tls::client","spans":[{"orig_dst":"10.2.0.179:8000","name":"ingress"},{"dst":"10.2.0.179:8000","name":"target"}],"threadId":"ThreadId(2)"}\n","stream":"stdout","time":"2021-05-13T14:36:12.108807334Z"}
2021-05-13T16:36:12+02:00 {} {"log":"{"timestamp":"[2066726.011443s]","level":"DEBUG","fields":{"message":"Connecting","server.addr":"10.2.0.179:8000"},"target":"linkerd_proxy_transport::connect","spans":[{"orig_dst":"10.2.0.179:8000","name":"ingress"},{"dst":"10.2.0.179:8000","name":"target"}],"threadId":"ThreadId(2)"}\n","stream":"stdout","time":"2021-05-13T14:36:12.108819514Z"}
2021-05-13T16:36:12+02:00 {} {"log":"{"timestamp":"[2066726.065720s]","level":"DEBUG","fields":{"message":"Failed to connect","error":"request timed out"},"target":"linkerd_reconnect::service","spans":[{"orig_dst":"10.2.6.171:8000","name":"ingress"},{"dst":"10.2.6.171:8000","name":"target"}],"threadId":"ThreadId(3)"}\n","stream":"stdout","time":"2021-05-13T14:36:12.163054753Z"}
```
#### `linkerd check` output

```text
kubernetes-api
--------------
√ can initialize the client
√ can query the Kubernetes API

kubernetes-version
------------------
√ is running the minimum Kubernetes API version
√ is running the minimum kubectl version

linkerd-existence
-----------------
√ 'linkerd-config' config map exists
√ heartbeat ServiceAccount exist
√ control plane replica sets are ready
√ no unschedulable pods
√ controller pod is running

linkerd-config
--------------
√ control plane Namespace exists
√ control plane ClusterRoles exist
√ control plane ClusterRoleBindings exist
√ control plane ServiceAccounts exist
√ control plane CustomResourceDefinitions exist
√ control plane MutatingWebhookConfigurations exist
√ control plane ValidatingWebhookConfigurations exist
√ control plane PodSecurityPolicies exist

linkerd-identity
----------------
√ certificate config is valid
√ trust anchors are using supported crypto algorithm
√ trust anchors are within their validity period
√ trust anchors are valid for at least 60 days
√ issuer cert is using supported crypto algorithm
√ issuer cert is within its validity period
‼ issuer cert is valid for at least 60 days
    issuer certificate will expire on 2021-06-18T15:46:36Z
    see l5d-identity-issuer-cert-not-expiring-soon for hints
√ issuer cert is issued by the trust anchor

linkerd-webhooks-and-apisvc-tls
-------------------------------
√ proxy-injector webhook has valid cert
‼ proxy-injector cert is valid for at least 60 days
    certificate will expire on 2021-06-18T15:46:29Z
    see l5d-proxy-injector-webhook-cert-not-expiring-soon for hints
√ sp-validator webhook has valid cert
‼ sp-validator cert is valid for at least 60 days
    certificate will expire on 2021-06-18T15:46:31Z
    see l5d-sp-validator-webhook-cert-not-expiring-soon for hints

linkerd-api
-----------
√ control plane pods are ready
√ can initialize the client
√ can query the control plane API

linkerd-version
---------------
√ can determine the latest version
‼ cli is up-to-date
    is running version 21.4.3 but the latest edge version is 21.4.5
    see l5d-version-cli for hints

control-plane-version
---------------------
‼ control plane is up-to-date
    is running version 21.4.3 but the latest edge version is 21.4.5
    see l5d-version-control for hints
√ control plane and cli versions match

Status check results are √
```

### Environment

- Kubernetes Version: 1.19.8
- Cluster Environment: (EKS)
- Host OS: Amazon Linux 2
- Linkerd version: edge-21.4.3

### Additional context

This bug report results from this [Slack discussion](https://linkerd.slack.com/archives/C89RTCWJF/p1620831632030300). The effects of this bug are quite similar to those of the fixed bug #5975.

**Comment by rustrial** [comment]
@olix0r Thanks for pointing me at #6037, I did not spot that one when scanning the release logs for bug fixes. Maybe it would help if you add links to the fixed issued and merged PRs to the release change log. 
We will now start testing `edge-21.5.1` and if not problems pop-up I will close this issue (as a duplicate of #6037).

**Comment by rustrial** [comment]
Closing as this is very likely a duplicate of #6037 which has already been fixed in edge-2.4.5.
